src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/PipelineJoiner.java

```diff
@@ -0,0 +1,182 @@
+    WaitHelper.waitForElementToBeDisplayed(Helper.locateElementByTestId(joinerFieldAlias + "-schema-field"));
+    Commands.closeConfigPopover();
+  }
+
+  @Then("Configure sink property and validate")
+  public void configureSinkNodeProperty() {
+    openNodeProperty(Commands.joinerSinkNode);
+    Commands.sendKeysToInputElementByTestId("referenceName", "BQ_Sink");
+    Commands.clearInputAndSendKeysToElementByTestId("project", Constants.DEFAULT_GCP_PROJECTID);
+    Commands.sendKeysToInputElementByTestId("dataset", joinerTestDataset);
+    Commands.sendKeysToInputElementByTestId("table", joinerTable3);
+    Commands.clearInputAndSendKeysToElementByTestId("serviceFilePath", Constants.DEFAULT_GCP_SERVICEACCOUNT_PATH);
+    WaitHelper.waitForElementToBeDisplayed(Helper.locateElementByTestId(joinerFieldAlias + "-schema-field"));
+    Commands.closeConfigPopover();
+  }
+
+  @Then("Joiner should still render Get Schema button when numPartitions is a macro")
+  public void joinerMacroValidation() {
+    openNodeProperty(Commands.joinerJoinerNode);
+    WaitHelper.waitForElementToBeDisplayed(Helper.locateElementByTestId("get-schema-btn"));
+    Commands.clearInputAndSendKeysToElementByTestId("numPartitions", "${testing.macro}");
+    Commands.closeConfigPopover();
+    openNodeProperty(Commands.joinerJoinerNode);
+    WaitHelper.waitForElementToBeDisplayed(Helper.locateElementByTestId("get-schema-btn"));
+    Commands.clearInputAndSendKeysToElementByTestId("numPartitions", "");
+    Commands.closeConfigPopover();
+  }
+
+  @Then("Should show appropriate message when preview has not been run yet")
+  public void beforePreviewRun() {
+    ElementHelper.clickOnElement(Helper.locateElementByTestId("pipeline-preview-btn"));
+    openPreviewPopover(Commands.joinerJoinerNode);
+    Helper.locateElementByXPath("//*[contains(text(),'Run preview to generate preview data')]");
+    Commands.closeConfigPopover();
+  }
+
+  @Then("Should show appropriate message when preview has been stopped before data is generated")
+  public void startAndStopPreview() {
+    ElementHelper.clickOnElement(Helper.locateElementByTestId("preview-top-run-btn"));
+    ElementHelper.clickOnElement(Helper.locateElementByTestId("stop-preview-btn"));
+    WaitHelper.waitForElementToBeDisplayed(Helper.locateElementByTestId("preview-top-run-btn"));
+    openPreviewPopover(Commands.joinerSinkNode);
+    Helper.locateElementByXPath("//*[contains(text(),'Input records have not been generated')]");
+    Commands.tryToCloseAlert();
+    Commands.closeConfigPopover();
+  }
+
+  @Then("Should show preview data with record view by default for sink")
+  public void startAndFinishPreview() {
+    ElementHelper.clickOnElement(Helper.locateElementByTestId("preview-top-run-btn"));
+    WaitHelper.waitForElementToBeDisplayed(Helper.locateElementByTestId("stop-preview-btn"));
+    //let it run for 60 seconds
+    Helper.waitSeconds(60);
```

Comment:
Is this additional wait needed?

Reply:
Yea, the preview takes about 1 min to finish run

Reply:
Generally, these wait method gives unpredictable results, better to avoid.  It may take more time if system is slow. Instead, we can wait for the element in subsequent screen.  
I think `WaitHelper.waitForElementToBeDisplayed(Helper.locateElementByTestId("preview-top-run-btn"));` would be enough here, can you please check once?

Reply:
I tried this but it didn't work. I thought it has a 180s timeout but it's not waiting. Not sure why

Reply:
ok. got it working with waitForElementToBeHidden

Reply:
great Sumeng !!